GREYErrorFormatter.m

```diff
@@ -0,0 +1,195 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "GREYErrorFormatter.h"
+
+#import "GREYObjectFormatter.h"
+#import "GREYInteraction.h"
+#import "GREYErrorConstants.h"
+#import "NSError+GREYCommon.h"
+#import "GREYError+Private.h"
+#import "GREYFatalAsserts.h"
+
+#pragma mark - String Constants
+
+/**
+ * All keys that can be added to the description logger in GREYErrorFormatter.
+ */
+static NSString *const kErrorFormatterExceptionReasonKey    = @"Exception Reason";
+static NSString *const kErrorFormatterRecoverySuggestionKey = @"Recovery Suggestion";
+static NSString *const kErrorFormatterElementMatcherKey     = @"Element Matcher";
+static NSString *const kErrorFormatterSearchActionInfoKey   = @"Search Action Info";
+static NSString *const kErrorFormatterScreenshotPathsKey    = @"Screenshot Paths";
+static NSString *const kErrorFormatterUnderlyingErrorKey    = @"Underlying Error";
+static NSString *const kErrorFormatterUIHierarchyKey        = @"UI Hierarchy";
+static NSString *const kErrorFormatterStackTraceKey         = @"Stack Trace";
+
+/**
+ * Keys used for logging the UI Hierarchy
+ */
+static NSString *const kHierarchyWindowLegendKey                 = @"[Window 1]";
+static NSString *const kHierarchyAcessibilityLegendKey           = @"[AX]";
+static NSString *const kHierarchyUserInteractionEnabledLegendKey = @"[UIE]";
+static NSString *const kHierarchyBackWindowKey                   = @"Back-Most Window";
+static NSString *const kHierarchyAccessibilityKey                = @"Accessibility";
+static NSString *const kHierarchyUserInteractionEnabledKey       = @"User Interaction Enabled";
+static NSString *const kHierarchyLegendKey                       = @"Legend";
+static NSString *const kHierarchyHeaderKey                       = @"UI Hierarchy (ordered by wind"
+                                                                   @"ow level, back to front):\n";
+
+#pragma mark - Private Variables
+
+@interface GREYErrorFormatter ()
+@property(readonly, nonatomic) GREYError *error;
+@end
+
+@implementation GREYErrorFormatter
+
+#pragma mark - Init
+
+- (instancetype)initWithError:(GREYError *)error {
+  self = [super init];
+  if (self) {
+    _error = error;
+  }
+  return self;
+}
+
+#pragma mark - Public Methods
+
+- (NSString *)formattedDescription {
+  if (GREYShouldUseErrorFormatterForError(_error)) {
+    return [self loggerDescriptionForKeys:[self loggerKeys]];
+  }
+  return [GREYObjectFormatter formatDictionary:[_error grey_descriptionDictionary]
+                                        indent:kGREYObjectFormatIndent
+                                     hideEmpty:YES
+                                      keyOrder:nil];
+}
+
+#pragma mark - Private Methods
+
+- (NSSet<NSString *> *)loggerKeys {
```

Comment:
Can we cache and reuse the same set? Instead of creating a new one every time.

Reply:
But won't it only be called once? Why would someone instantiate an ErrorFormatter and repeatedly call this method?

Reply:
They can do it when repeatedly performing an interaction and checking the error description. This is only called internally in this function - this can be made a static function as well.

Reply:
Fixed